File: cpp/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.empty()) return nullptr;

        const unsigned int size = nums.size() / 2;
        TreeNode* root = new TreeNode(nums[size]);

        vector<int> left(nums.begin(), nums.begin() + nums.size() / 2);
        vector<int> right(nums.begin() + nums.size() / 2 + 1, nums.end());

        root->left = sortedArrayToBST(left);
        root->right = sortedArrayToBST(right);

        return root;
    }
};

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) { return nullptr; }

        queue<TreeNode*> node_queue;
        int count = 0;

        TreeNode* root = new TreeNode(0);
        TreeNode* current = nullptr;

        node_queue.push(root);
        ++count;

        while (count < nums.size()) {
            current = node_queue.front();
            node_queue.pop();
            current->left = new TreeNode(0);
            ++count;
            node_queue.push(current->left);
            
            if (count < nums.size()) {
                current->right = new TreeNode(0);
                ++count;
                node_queue.push(current->right);
            }
        }
        int i = 0;
        populate(root, nums, i);
        return root;
        
    }
    
    void populate(TreeNode* root, vector<int>& nums, int &i){
        if (!root) { return; }
        populate(root->left, nums, i);
        root->val=nums[i];
        ++i;
        populate(root->right, nums, i);
    }
    
    
};

File: py/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# Recursive solution, O(n) for time and space since we need to look at every
# node at least once and store every node to construct the tree
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return None
        mid = len(nums) // 2
        low = nums[:mid]
        high = nums[mid+1:]
        root = TreeNode(nums[mid])
        root.left = self.sortedArrayToBST(low)
        root.right = self.sortedArrayToBST(high)
        return root

# Helper function to recurse approach
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.bst_helper(nums, 0, len(nums))

    def bst_helper(self, nums, left, right):
        if left >= right:
            return None

        # mid = len(nums) // 2 #
        mid = left + (right - left) // 2
        root = TreeNode(nums[mid])
        root.left = self.bst_helper(nums, left, mid)
        root.right = self.bst_helper(nums, mid+1, right)
        return root

# Ecapsulated helper function
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        
        def bst_helper(left, right):
            if left > right:
                return None
            
            mid = (left + right) // 2
            root = TreeNode(nums[mid])
            
            root.left = bst_helper(left, mid-1)
            root.right = bst_helper(mid+1, right)
            
            return root
        
    
        return bst_helper(0, len(nums) - 1)